### PR TITLE
Tweak riscv's cpu_fetch_syscall_args() a bit.

### PR DESCRIPTION
**Pull request by trasz**
Mostly to make it possible to inline memcpy.

### CODE REVIEW

File: sys/riscv/riscv/trap.c
Comment (arichardson) [inline sys/riscv/riscv/trap.c:141]: You can avoid a branch by copying the first value here and incrementing the destination pointer and then copying NARGS - 1 below:

https://cheri-compiler-explorer.cl.cam.ac.uk/z/T4fdEa

Comment (trasz) [reply sys/riscv/riscv/trap.c]: Nice, thanks!

**Review by arichardson — APPROVED** [review]
LGTM, but maybe wait for one more review?

File: sys/riscv/riscv/trap.c
Comment (jrtc27) [inline sys/riscv/riscv/trap.c:141]: Why not:
```suggestion
		*dst_ap++ = *ap++;
```
? Seems weird to increment one but not the other like that.

File: sys/riscv/riscv/trap.c
Comment (brooksdavis) [inline sys/riscv/riscv/trap.c:150]:
```suggestion
	KASSERT(sa->callp->sy_narg <= nitems(sa->args), ("Syscall %d takes too many arguments!", sa->code));
```
That being said, this probably ought to be verified when the sysent is installed and when modules install syscalls rather than in this path.

Comment (brooksdavis) [reply sys/riscv/riscv/trap.c:150]: Also, in the syscall() and __syscall() cases this needs to be `nitems(sa->args) - 1`.

**Review by brooksdavis — APPROVED** [review]
Other than some nits with the assert I think this looks fine.

**Comment by jrtc27** [comment]
Should this not be approved upstream then cherry-picked back? The upstream code uses a variable for the `memcpy` too, it's not just down to the local handling of `SYS_syscall`/`SYS___syscall`. Also, this applies just as equally to arm64, where the original code was presumably copied from.

File: sys/riscv/riscv/trap.c
Comment (trasz) [reply sys/riscv/riscv/trap.c:150]: Erm... why?  In case of syscall() and __syscall() the arguments are shifted in registers, but in sa->args they are in their normal order, ie the same as if the "target syscall" was called directly, without going through syscall(2).

**Comment by trasz** [comment]
Upstream review: https://reviews.freebsd.org/D26528

**Comment by arichardson** [comment]
I guess this can be closed since @brooksdavis is going to do an upstream merge?

**Comment by brooksdavis** [comment]
Merged from upstream in 308955e31c4eb48d9143834feb1deca10dcc3637

**Pull request closed without merging**
